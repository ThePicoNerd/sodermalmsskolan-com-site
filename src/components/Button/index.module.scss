.base {
  background-color: transparent;
  border: 0;
  border-radius: 6px;
  box-sizing: border-box;
  color: var(--color-text-primary);
  cursor: pointer;
  font-family: inherit;
  font-size: 14px;
  font-weight: 500;
  letter-spacing: 0;
  transition: background-color 0.1s;
}

.primary {
  background-color: var(--color-highlight);
  color: #fff;

  &:hover {
    background-color: var(--color-highlight-light);
  }

  &:active {
    background-color: var(--color-highlight);
  }
}

.secondary {
  border: 1px solid var(--accents-2);

  &:hover {
    background-color: var(--accents-1);
  }

  &:active {
    background-color: var(--accents-2);
  }
}

.small {
  padding: 0.5rem 1rem;
}

.medium {
  padding: 0.75rem 1rem;
}

.large {
  padding: 1rem 1.5rem;
}
