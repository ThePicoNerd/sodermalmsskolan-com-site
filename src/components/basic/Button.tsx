import styled from "styled-components";
import React from "react";
import Link from "next/link";
import * as breakpoints from "../../styles/breakpoints";
import { lighten, transparentize } from "polished";
import { motion } from "framer-motion";

const ButtonIcon = styled.span`
  display: inline-flex;
  align-items: center;
  vertical-align: middle;
  margin-top: -4px;
  margin-left: 1em;
  transition: transform 0.2s ease-in-out;

  svg {
    height: 1.5rem;
    width: 1.5rem;
  }
`;

export const IconButton = styled(motion.a)<{ $disabled?: boolean }>`
  background-color: transparent;
  color: ${({ theme, $disabled }) =>
    $disabled ? theme.colors.border : theme.colors.primary};
  display: inline-block;
  padding: 0.5rem;
  border-radius: 50%;
  cursor: ${({ $disabled }) => ($disabled ? "not-allowed" : "pointer")};
  transition: all 0.1s ease;

  &:hover {
    background-color: ${({ theme, $disabled }) =>
      !$disabled &&
      transparentize(theme.dark ? 0.75 : 0.9, theme.colors.primary)};
  }

  svg {
    height: 1.5em;
    width: 1.5em;
    display: block;
  }
`;

export const StyledButton = styled.a<{
  $secondary?: boolean;
  $background?: string;
  $foreground?: string;
  $height: number;
}>`
  display: inline-block;
  box-sizing: border-box;
  background-color: ${({ theme, $background }) =>
    $background || theme.colors.primary};
  color: ${({ theme, $foreground }) => $foreground || theme.colors.background};
  padding: 0 ${({ $height }) => `${$height / 2}rem`};
  border-radius: ${({ $height }) => `${$height / 2}rem`};
  line-height: ${({ $height }) => `${$height}rem`};
  transition: background-color 0.2s ease, color 0.2s ease;
  cursor: pointer;

  &:hover {
    background-color: ${({ theme, $background }) =>
      lighten(0.1, $background || theme.colors.primary)};

    ${ButtonIcon} {
      transform: translateX(4px);
    }
  }

  ${({ $secondary, theme, $foreground, $background }) =>
    $secondary &&
    `
    background-color: ${transparentize(
      theme.dark ? 0.75 : 0.9,
      $background || theme.colors.primary
    )};
    color: ${$foreground || theme.colors.foreground};

    &:hover {
      background-color: ${transparentize(
        theme.dark ? 0.5 : 0.75,
        $background || theme.colors.primary
      )};
    }
  `}
`;

export interface ButtonProps extends React.HTMLAttributes<HTMLAnchorElement> {
  secondary?: boolean;
  large?: boolean;
  href?: string;
  foreground?: string;
  background?: string;
  icon?: React.ReactNode;
}

export const Button: React.FunctionComponent<ButtonProps> = ({
  secondary,
  large,
  href,
  children,
  foreground,
  background,
  icon,
  ...rest
}) => {
  const height = large ? 3.5 : 3;

  const button = (
    <StyledButton
      $secondary={secondary}
      $height={height}
      $foreground={foreground}
      $background={background}
      {...rest}
    >
      {children}
      {icon && <ButtonIcon>{icon}</ButtonIcon>}
    </StyledButton>
  );

  if (href) {
    return (
      <Link href={href} passHref>
        {button}
      </Link>
    );
  } else {
    return button;
  }
};

export const ButtonRow = styled.div<{ center?: boolean }>`
  display: inline-flex;
  flex-wrap: wrap;
  margin: -12px;
  justify-content: center;

  @media (min-width: ${breakpoints.large}) {
    justify-content: normal;
  }

  ${({ center }) =>
    center &&
    `
    justify-content: center;
  `}

  a {
    margin: 12px;
  }
`;
