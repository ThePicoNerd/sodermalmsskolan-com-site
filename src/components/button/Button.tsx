/* eslint-disable jsx-a11y/anchor-has-content */
import React, { FunctionComponent } from "react";
import Link from "next/link";
import { css, CSSObject, SerializedStyles } from "@emotion/react";
import { fonts } from "../../styles/text";

export type ButtonSize = "medium" | "small";

export interface ButtonProps {
  /**
   * Whether or not the button is primary.
   */
  primary?: boolean;
  href?: string;
  size?: ButtonSize;
  onClick?: () => void;
  disabled?: boolean;
  type?: "button" | "submit" | "reset",
}

const disabledSelector = "&:disabled, &[disabled]";

/**
 * Base styles of a button. Not to be used on its own.
 *
 * @param {ButtonSize} size The desired size of the button.
 *
 * @returns {SerializedStyles} Styles.
 */
const baseStyles = (size: ButtonSize): SerializedStyles => {
  let paddingX: string;
  let paddingY: string;

  // eslint-disable-next-line default-case
  switch (size) {
    case "medium": {
      paddingX = "1.5rem";
      paddingY = "1rem";
      break;
    }

    case "small": {
      paddingX = "1rem";
      paddingY = "0.75rem";
      break;
    }
  }

  return css({
    color: "var(--color-text-primary)",
    backgroundColor: "transparent",
    display: "inline-block",
    border: "none",
    outline: "none",
    fontFamily: fonts.sans,
    fontWeight: 600,
    fontSize: "1rem",
    whiteSpace: "nowrap",
    borderRadius: "0.375rem",
    cursor: "pointer",
    textDecoration: "none",
    transition: "all 0.1s",
    padding: `${paddingY} ${paddingX}`,
    lineHeight: 1,

    [disabledSelector]: {
      boxShadow: "inset 0 0 0 1px var(--accents-2)",
      color: "transparent",
      pointerEvents: "none",
    },
  });
};

const primaryStyles = css({
  backgroundColor: "var(--color-highlight)",
  backgroundImage: "linear-gradient(180deg, var(--color-highlight-light), var(--color-highlight))",
  color: "#ffffff",

  "&:hover": {
    opacity: 0.75,
  },

  [disabledSelector]: {
    background: "var(--accents-1)",
  },
});

const secondaryStyles = css({
  boxShadow: "inset 0 0 0 1px var(--color-text-secondary)",
  transition: "box-shadow 0.2s ease",

  "&:hover": {
    boxShadow: "inset 0 0 0 2px var(--color-text-primary)",
  },
});

/**
 * A button with a link. Neat, right?
 *
 * @param {React.PropsWithChildren<ButtonProps>} props Button props.
 * @param {string} props.href Optional `href`.
 * @param {boolean} props.primary Is the button primary?
 * @param {ButtonSize} props.size Size of the button.
 * @param {boolean} props.disabled Is the button clickable?
 *
 * @returns {React.ReactElement} The rendered button.
 */
const Button: FunctionComponent<ButtonProps> = ({
  href,
  primary = false,
  size = "medium",
  disabled = false,
  ...rest
}) => {
  const styles: (SerializedStyles | CSSObject)[] = [
    baseStyles(size),
    primary ? primaryStyles : secondaryStyles,
  ];

  const innerProps = {
    css: styles,
    disabled,
    ...rest,
  };

  if (href) {
    return (
      <Link href={href} passHref>
        <a {...innerProps} />
      </Link>
    );
  }
  // eslint-disable-next-line react/button-has-type
  return <button {...innerProps} />;
};

export default Button;
