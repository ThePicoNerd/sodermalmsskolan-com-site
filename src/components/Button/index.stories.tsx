import React, { ComponentProps } from "react";

// eslint-disable-next-line import/no-extraneous-dependencies
import { Story } from "@storybook/react";
import { ArrowRight, Zap } from "react-feather";
import Button from ".";

export default {
  title: "Button",
  component: Button,
  argTypes: {
    icon: {
      table: {
        disable: true,
      },
    },
    leftIcon: {
      table: {
        disable: true,
      },
    },
    rightIcon: {
      table: {
        disable: true,
      },
    },
  },
};

// eslint-disable-next-line require-jsdoc
const Template: Story<ComponentProps<typeof Button>> = (args) => <Button {...args} />;

export const Primary = Template.bind({});

Primary.args = {
  children: "Primary",
  variant: "primary",
};

export const Secondary = Template.bind({});

Secondary.args = {
  children: "Secondary",
  variant: "secondary",
};

export const Tertiary = Template.bind({});

Tertiary.args = {
  children: "Tertiary",
  variant: "tertiary",
};

export const Disabled = Template.bind({});

Disabled.args = {
  children: "Disabled",
  disabled: true,
};

export const WithIcon = Template.bind({});

WithIcon.args = {
  children: "Continue",
  variant: "primary",
  rightIcon: ArrowRight,
};

export const Icon = Template.bind({});

Icon.args = {
  icon: Zap,
};
