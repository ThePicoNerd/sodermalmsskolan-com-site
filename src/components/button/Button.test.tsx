import renderer from "react-test-renderer";
import React from "react";
import Button from "./Button";

describe("button test", () => {
  it("should render primary buttons correctly", () => {
    const tree = renderer
      .create(<Button primary>My primary button</Button>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render small buttons correctly", () => {
    const tree = renderer
      .create(<Button size="small">My small button</Button>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
