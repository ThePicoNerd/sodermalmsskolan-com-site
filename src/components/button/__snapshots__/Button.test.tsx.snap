// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> test should render primary buttons correctly 1`] = `
.emotion-0 {
  color: var(--color-text-primary);
  background-color: transparent;
  display: inline-block;
  border: none;
  outline: none;
  font-family: Inter,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Oxygen,Ubuntu,Cantarell,"Open Sans","Helvetica Neue",Helvetica,sans-serif;
  font-weight: 600;
  font-size: 1rem;
  white-space: nowrap;
  border-radius: 0.375rem;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: all 0.1s;
  transition: all 0.1s;
  padding: 1rem 1.5rem;
  line-height: 1;
  background-color: var(--color-highlight);
  background-image: linear-gradient(180deg, var(--color-highlight-light), var(--color-highlight));
  color: #ffffff;
}

.emotion-0:disabled,
.emotion-0[disabled] {
  box-shadow: inset 0 0 0 1px var(--accents-2);
  color: transparent;
  pointer-events: none;
}

.emotion-0:hover {
  opacity: 0.75;
}

.emotion-0:disabled,
.emotion-0[disabled] {
  background: var(--accents-1);
}

<button
  className="emotion-0"
  disabled={false}
>
  My primary button
</button>
`;

exports[`<Button /> test should render small buttons correctly 1`] = `
.emotion-0 {
  color: var(--color-text-primary);
  background-color: transparent;
  display: inline-block;
  border: none;
  outline: none;
  font-family: Inter,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Oxygen,Ubuntu,Cantarell,"Open Sans","Helvetica Neue",Helvetica,sans-serif;
  font-weight: 600;
  font-size: 1rem;
  white-space: nowrap;
  border-radius: 0.375rem;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: all 0.1s;
  transition: all 0.1s;
  padding: 0.75rem 1rem;
  line-height: 1;
  box-shadow: inset 0 0 0 1px var(--color-text-secondary);
  -webkit-transition: box-shadow 0.2s ease;
  transition: box-shadow 0.2s ease;
}

.emotion-0:disabled,
.emotion-0[disabled] {
  box-shadow: inset 0 0 0 1px var(--accents-2);
  color: transparent;
  pointer-events: none;
}

.emotion-0:hover {
  box-shadow: inset 0 0 0 2px var(--color-text-primary);
}

<button
  className="emotion-0"
  disabled={false}
>
  My small button
</button>
`;
